k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -0,0 +1,308 @@
+package io.buoyant.k8s
+
+import com.twitter.concurrent.AsyncStream
+import com.twitter.finagle.http.Response
+import com.twitter.finagle.{Filter, http}
+import com.twitter.finagle.param.HighResTimer
+import com.twitter.finagle.service.{Backoff, RetryBudget, RetryFilter, RetryPolicy}
+import com.twitter.finagle.stats.{DefaultStatsReceiver, StatsReceiver}
+import com.twitter.io.{Buf, Reader}
+import com.twitter.util.{Closable, _}
+import com.twitter.util.TimeConversions._
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Basic functionality shared by all k8s Resource implementations: they require an HTTP client, and can be closed.
+ */
+trait Resource extends Closable {
+  def client: Client
+  def close(deadline: Time) = client.close(deadline)
+}
+
+/**
+ * A Version is a top-level API grouping within the K8s API; examples include "/api/v1" and "/apis/extensions/v1beta1".
+ *
+ * Implementors are likely to want to extend this to provide named convenience methods on top of the `list` method and
+ * to type-specialize on the resource types supported by that version.
+ *
+ * @tparam O The parent type for all [Object]s served by this version.
+ */
+private[k8s] trait Version[O <: Object] extends Resource {
```

Comment:
so I'm trying to figure out the right name scheme. Right now, the Kubernetes API for `apis` returns an `ApiGroupList`, with items like:
```
{
      "name": "buoyant.io",
      "versions": [
        {
          "groupVersion": "buoyant.io/v1",
          "version": "v1"
        }
      ],
    },
```

and "/api" returns an `ApiVersionList`, of which the only element is "v1".

So: it sounds like the closest nomenclature to what we want is actually "GroupVersion"? Or we could introduce another layer in our API, so you'd have something like Group("buoyant").version("v1").namespace("ns")...?